metrics/prometheus/prometheus_test.go

```diff
@@ -8,80 +8,121 @@ import (
+	register := func(t *testing.T, name string, collector any) {
+		t.Helper()
+		err := registry.Register(name, collector)
+		require.NoError(t, err)
+	}
 
 	counter := metrics.NewCounter()
 	counter.Inc(12345)
+	register(t, "test/counter", counter)
 
-	err := registry.Register("test/counter", counter)
-	assert.NoError(t, err)
+	counterFloat64 := metrics.NewCounterFloat64()
+	counterFloat64.Inc(1.1)
+	register(t, "test/counter_float64", counterFloat64)
 
 	gauge := metrics.NewGauge()
 	gauge.Update(23456)
-
-	err = registry.Register("test/gauge", gauge)
-	assert.NoError(t, err)
+	register(t, "test/gauge", gauge)
 
 	gaugeFloat64 := metrics.NewGaugeFloat64()
 	gaugeFloat64.Update(34567.89)
-
-	err = registry.Register("test/gauge_float64", gaugeFloat64)
-	assert.NoError(t, err)
+	register(t, "test/gauge_float64", gaugeFloat64)
 
 	sample := metrics.NewUniformSample(1028)
 	histogram := metrics.NewHistogram(sample)
-
-	err = registry.Register("test/histogram", histogram)
-	assert.NoError(t, err)
+	register(t, "test/histogram", histogram)
 
 	meter := metrics.NewMeter()
-	defer meter.Stop()
+	t.Cleanup(meter.Stop)
 	meter.Mark(9999999)
-
-	err = registry.Register("test/meter", meter)
-	assert.NoError(t, err)
+	register(t, "test/meter", meter)
 
 	timer := metrics.NewTimer()
-	defer timer.Stop()
+	t.Cleanup(timer.Stop)
 	timer.Update(20 * time.Millisecond)
 	timer.Update(21 * time.Millisecond)
 	timer.Update(22 * time.Millisecond)
 	timer.Update(120 * time.Millisecond)
 	timer.Update(23 * time.Millisecond)
 	timer.Update(24 * time.Millisecond)
-
-	err = registry.Register("test/timer", timer)
-	assert.NoError(t, err)
+	register(t, "test/timer", timer)
 
 	resettingTimer := metrics.NewResettingTimer()
-	resettingTimer.Update(10 * time.Millisecond)
-	resettingTimer.Update(11 * time.Millisecond)
-	resettingTimer.Update(12 * time.Millisecond)
-	resettingTimer.Update(120 * time.Millisecond)
-	resettingTimer.Update(13 * time.Millisecond)
-	resettingTimer.Update(14 * time.Millisecond)
-
-	err = registry.Register("test/resetting_timer", resettingTimer)
-	assert.NoError(t, err)
-
-	err = registry.Register("test/resetting_timer_snapshot", resettingTimer.Snapshot())
-	assert.NoError(t, err)
+	register(t, "test/resetting_timer", resettingTimer)
+	resettingTimer.Update(time.Second) // must be after register call
 
 	emptyResettingTimer := metrics.NewResettingTimer()
-
-	err = registry.Register("test/empty_resetting_timer", emptyResettingTimer)
-	assert.NoError(t, err)
-
-	err = registry.Register("test/empty_resetting_timer_snapshot", emptyResettingTimer.Snapshot())
-	assert.NoError(t, err)
-
-	g := Gatherer(registry)
-
-	_, err = g.Gather()
-	assert.NoError(t, err)
+	register(t, "test/empty_resetting_timer", emptyResettingTimer)
+
+	emptyResettingTimer.Update(time.Second) // no effect because of snapshot below
+	register(t, "test/empty_resetting_timer_snapshot", emptyResettingTimer.Snapshot())
+
+	// Nil metrics
+	metrics.Enabled = false
```

Comment:
Why does this need to be done with metrics disabled?

Reply:
Because nil metrics are only built when Enabled is false.

For example `NilCounter` is returned by `NewCounter()` only if `Enabled` is `false`.

Reply:
I learnt that later in the day. Not a fan of the API.